Mathlib/Algebra/Associated.lean

```diff
@@ -0,0 +1,1262 @@
+  exact ha.1.not_unit ha.2
+#align irreducible.not_square Irreducible.not_square
+
+theorem IsSquare.not_irreducible (ha : IsSquare a) : ¬Irreducible a := fun h => h.not_square ha
+#align is_square.not_irreducible IsSquare.not_irreducible
+
+end CommMonoid
+
+section CancelCommMonoidWithZero
+
+variable [CancelCommMonoidWithZero α] {a p : α}
+
+protected theorem Prime.irreducible (hp : Prime p) : Irreducible p :=
+  ⟨hp.not_unit, fun a b hab =>
+    (show a * b ∣ a ∨ a * b ∣ b from hab ▸ hp.dvd_or_dvd (hab ▸ dvd_rfl)).elim
+      (fun ⟨x, hx⟩ =>
+        Or.inr
+          (isUnit_iff_dvd_one.2
+            ⟨x,
+              mul_right_cancel₀ (show a ≠ 0 from fun h => by
+                simp [Prime] at *
+                rw [h, zero_mul] at hab
+                have := hp.left
+                contradiction
+                ) <| by
+                conv =>
+                    lhs
+                    rw [hx]
+                · simp [mul_comm, mul_assoc, mul_left_comm]
+                ⟩))
+      fun ⟨x, hx⟩ =>
+      Or.inl
+        (isUnit_iff_dvd_one.2
+          ⟨x,
+            mul_right_cancel₀ (show b ≠ 0 from fun h => by
+            simp [Prime] at *
+            rw [h, mul_zero] at hab
+            have := hp.left
+            contradiction
+            ) <| by
+              conv =>
+                  lhs
+                  rw [hx]
+              · simp [mul_comm, mul_assoc, mul_left_comm]⟩)⟩
+#align prime.irreducible Prime.irreducible
+
+theorem succ_dvd_or_succ_dvd_of_succ_sum_dvd_mul (hp : Prime p) {a b : α} {k l : ℕ} :
+    p ^ k ∣ a → p ^ l ∣ b → p ^ (k + l + 1) ∣ a * b → p ^ (k + 1) ∣ a ∨ p ^ (l + 1) ∣ b :=
+  fun ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩ =>
+  have h : p ^ (k + l) * (x * y) = p ^ (k + l) * (p * z) := by
+    simpa [mul_comm, pow_add, hx, hy, mul_assoc, mul_left_comm] using hz
+  have hp0 : p ^ (k + l) ≠ 0 := pow_ne_zero _ hp.ne_zero
+  have hpd : p ∣ x * y := ⟨z, by rwa [mul_right_inj' hp0] at h⟩
+  (hp.dvd_or_dvd hpd).elim
+    (fun ⟨d, hd⟩ => Or.inl ⟨d, by simp [*, pow_succ, mul_comm, mul_left_comm, mul_assoc]⟩)
+    fun ⟨d, hd⟩ => Or.inr ⟨d, by simp [*, pow_succ, mul_comm, mul_left_comm, mul_assoc]⟩
+#align succ_dvd_or_succ_dvd_of_succ_sum_dvd_mul succ_dvd_or_succ_dvd_of_succ_sum_dvd_mul
+
+theorem Prime.not_square (hp : Prime p) : ¬IsSquare p :=
+  hp.irreducible.not_square
+#align prime.not_square Prime.not_square
+
+theorem IsSquare.not_prime (ha : IsSquare a) : ¬Prime a := fun h => h.not_square ha
+#align is_square.not_prime IsSquare.not_prime
+
+theorem pow_not_prime {n : ℕ} (hn : n ≠ 1) : ¬Prime (a ^ n) := fun hp =>
+  hp.not_unit <| IsUnit.pow _ <| of_irreducible_pow hn <| hp.irreducible
+#align pow_not_prime pow_not_prime
+
+end CancelCommMonoidWithZero
+
+/-- Two elements of a `monoid` are `associated` if one of them is another one
+multiplied by a unit on the right. -/
+def Associated [Monoid α] (x y : α) : Prop :=
+  ∃ u : αˣ, x * u = y
+#align associated Associated
+
+-- mathport name: «expr ~ᵤ »
```

Comment:
```suggestion
```

Reply:
This was from mathport. Should it be deleted?